src/internal/modules/invoice-accounts/lib/helpers.js

```diff
@@ -54,18 +60,47 @@ const getSelectedAddress = async (companyId, session) => {
   if (session.address.addressId === tempId) {
     return session.address;
   } else {
-    const addresses = await dataService.getCompanyAddresses(companyId);
+    const addresses = await getAllAddresses(companyId, session);
     const selectedAddress = addresses.find(address => (address.id === session.address.addressId));
     return selectedAddress;
   };
 };
 
-const getAgentCompany = (session) => {
-  if (has(session, 'agent')) {
-    return session.agent.companyId === tempId ? session.agent : dataService.getCompany(session.agent.companyId);
+const getAllAddresses = async (companyId, session) => {
+  let originalCompanyAddresses = await dataService.getCompanyAddresses(companyId) || [];
+  let agentCompanyAddresses = has(session, 'agent.companyId') && session.agent.companyId !== '00000000-0000-0000-0000-000000000000' ? await dataService.getCompanyAddresses(session.agent.companyId) : [] || [];
+  let newAddressFromSession = {};
+
+  if (has(session, 'address.id')) {
+    if (session.address.addressId === tempId) {
+      newAddressFromSession = session.address;
+    }
+  }
+
+  const compiledArrays = [...originalCompanyAddresses, ...agentCompanyAddresses, newAddressFromSession];
+  const filteredArray = compiledArrays.filter(obj => obj.id);
+  return filteredArray;
+};
+
+const getAgentCompany = async (session) => {
+  if (session.agent && session.agent.companyId) {
+    const outcome = session.agent.companyId === tempId ? session.agent : await dataService.getCompany(session.agent.companyId);
+    return outcome;
   } else { return null; }
 };
 
+const getCompanyName = async (request) => {
+  const { sessionKey } = request.query;
+  const { companyId } = request.params;
+  let currentState = await request.yar.get(sessionKey);
+  if (currentState.newCompany) {
+    return currentState.agent.name;
+  } else {
+    const originalCompany = await dataService.getCompany(companyId);
```

Comment:
this company name is already in the session viewData object

Reply:
Not necessarily - the company name in `session.viewData` may be over-written with the name of the agent company as part of the contact entry handover. :)